packages/integration-test-app/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -85,70 +81,81 @@ void TestHostHarness::OnInstanceLoaded(const InstanceLoadedEventArgs &args) noex
 
-  auto listenResult = co_await m_commandListener->StartListening();
+  auto cancellationToken = co_await winrt::get_cancellation_token();
 
-  switch (listenResult) {
-    case ListenResult::Success:
-      co_return;
+  // Keep on polling for clients until we die
+  while (!cancellationToken()) {
+    auto processReqsTask = m_rpcServer.ProcessAllClientRequests(8305 /*port*/, 50ms /*pollInterval*/);
 
-    case ListenResult::AddressInUse:
-      ShowJSError("Test harness address is already in use. Is there another instance running?");
-      co_return;
+    // Foward cancellation to also cancel processing reqs (setting callback replaces the last)
+    cancellationToken.callback([processReqsTask]() noexcept { processReqsTask.Cancel(); });
+    co_await processReqsTask;
+  }
+}
 
-    case ListenResult::OtherError:
-      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
-      co_return;
+winrt::NodeRpc::Handler TestHostHarness::CreateRpcHander() noexcept {
+  winrt::NodeRpc::Handler handler;
 
-    default:
-      ShowJSError("Unexpected ListenResult from TestCommandListener.");
-      co_return;
-  }
+  handler.BindAsyncOperation(
+      L"RunTestComponent", [weakThis{get_weak()}](const JsonValue &payload) noexcept -> IAsyncOperation<IJsonValue> {
+        auto strongThis = weakThis.get();
+        VerifyElseCrash(strongThis);
+        co_return co_await strongThis->OnTestCommand(TestCommandId::RunTestComponent, payload);
+      });
+
+  handler.BindAsyncOperation(
+      L"GoToComponent", [weakThis{get_weak()}](const JsonValue &payload) noexcept -> IAsyncOperation<IJsonValue> {
+        auto strongThis = weakThis.get();
+        VerifyElseCrash(strongThis);
+        co_return co_await strongThis->OnTestCommand(TestCommandId::GoToComponent, payload);
+      });
+
+  return handler;
 }
 
-winrt::fire_and_forget TestHostHarness::OnTestCommand(TestCommand command, TestCommandResponse response) noexcept {
+IAsyncOperation<IJsonValue> TestHostHarness::OnTestCommand(TestCommandId command, JsonValue payload) noexcept {
   VerifyElseCrash(m_dispatcher.HasThreadAccess());
 
   // Keep ourselves alive while we have a test command
   auto strongThis = get_strong();
 
   if (m_pendingResponse) {
-    response.Error("Received a test command while still processing the previous");
-    co_return;
+    co_return TestCommandResponse::Error("Received a test command while still processing the previous");
   }
 
   if (m_instanceFailedToLoad) {
-    response.Error("The instance failed to load");
-    co_return;
+    co_return TestCommandResponse::Error("The instance failed to load");
   }
 
-  m_pendingResponse = std::move(response);
+  m_pendingResponse.emplace();
 
-  switch (command.Id) {
+  switch (command) {
     case TestCommandId::RunTestComponent: {
-      auto componentName = command.payload.GetString();
+      auto componentName = payload.GetObject().GetNamedString(L"component");
```

Comment:
>component [](start = 64, length = 9)

Hmm, didn't this get renamed to the "name" field in `IntegrationTestRunner
.ts`

Reply:
Rename was "command" to "name", That should be reflected in native as well.

Though I'm not entirely sure it's the right name. "name" makes sense on "Command" objects, but is not self-describing within the JSON itself.